src/OrthoTree/OrthoTree.h

```diff
@@ -1,12 +1,167 @@
+         * - contents of tree_m
+         *
+         * @warning THIS FUNCTION ASSUMES THAT THE TREES ARE SORTED
+         *
+         * @param other
+         * @return true
+         * @return false
+         */
+        bool operator==(const OrthoTree& other);
+
+        /**
+         * @brief Returns a vector with morton_codes and lists of particle ids inside this region
+         * This is also not meant to be permanent, but it should suffice for the beginning
+         * @return Kokkos::vector<Kokkos::pair<morton_code, Kokkos::vector<size_t>>>
+         */
+        Kokkos::vector<Kokkos::pair<morton_code, Kokkos::vector<size_t>>> get_tree() const;
+
+    private:
+
+        /**
+         * @brief initializes the aid list which has form vec{pair{morton_code, size_t}}
+         * will sort the aidlist in ascending morton codes
+         *
+         * @param particles
+         */
+        void initialize_aid_list(particle_t const& particles);
+
+    public:
+
+        // SIMONS FUNCTIONS DONT EDIT, TOUCH OR USE THIS IN YOUR CODE:
+
+        /**
+         * @brief algorithm 1' topdown sequential construction of octree
+         *
+         * @param morton code of root_node, particles
+         *
+         * @return list of morton codes of leaves of tree spanning root node
+         **/
+        ippl::vector_t<morton_code> build_tree_topdown_sequential(morton_code root_node, particle_t const& particles)
+        {
+            // insert the root into the tree
+            ippl::vector_t<morton_code> tree;
+            ippl::vector_t<morton_code> stack; // stack used to build the tree
+
+            stack.push_back(root_node); // maybe store morton_code(0) s.t. we can call morton::root_val or smth
+
+            initialize_aid_list(particles); // initialize aid list - required for particle counting could be moved to constructor
+
+            while ( stack.size() > 0 ) {
+                morton_code current_node = stack.back();
+                stack.pop_back();
+
+                if ( NumberOfParticles(current_node) > max_particles_per_node_m && morton_helper.get_depth(current_node) < max_depth_m ) {
+
+                    ippl::vector_t<morton_code> children = morton_helper.get_children(current_node);
+
+                    for ( morton_code child : children ) { stack.push_back(child); }
+
+                }
+                else {
+                    tree.push_back(current_node);
+                }
+            }
+
+            return tree;
+        }
+
+        /**
+         * @brief counts the number of particles covered by the cell decribed by the morton code
+         *        initialize_aid_list needs to be called first
+         *
+         * @params morton_code
+         *
+         * @return number of particles in the cell specified by the morton code
+         *
+         **/
+        size_t NumberOfParticles(morton_code mc)
+        {
+            size_t count = 0;
+            for ( auto aid_list_el : aid_list ) {
```

Comment:
For a sorted aid_list the particles inside should be contiguous in aid_list so an efficient implementation is possible searching for lower/upper bounds. $\mathcal{O}(\log(n_p))$

Reply:
we are now using std::lower/upper_bound, runtime is now O(log(n))